### ISSUE TITLE
Test using object store on Jetstream

### ISSUE THREAD

**Issue by zonca** [MIRC-UofSC/xsede-development#1]
@junzhoum 

Here is what I need to get started on this:

* 1 test video not too big (~tens of MB)
* a self-contained example of a webpage that does streaming, ideally if you can extract it from your gateway prototype and simplify it it would be great. Whatever language you are using is fine.
* It would be ideal if this both does the streaming and also satisfies your security requirements (prevent downloading)
* You can commit the example to https://github.com/MIRC-UofSC/xsede-development, if not too big, you can also post the test video there

Then I'll upload the video to object store and try to adapt your example to stream from object store while still maintaining your security requirements.

**Comment by zonca** [comment]
ok, so sources are mp4 files and we want to prevent users from downloading them.
If we are deploying on Jetstream, then we do not have anything similar to Cloudfront streaming.
We need to deploy a webRTC server in a Jetstream virtual machine.

The idea would be to have the server retrieve the MP4 from Jetstream object store (very similar to S3) and then stream it piecewise to the user.

I have done a bit of research and found 2 options:

* A module for NGINX https://github.com/kaltura/nginx-vod-module, it allows `nginx` to get a file remotely from HTTP (this will be object store so over local network is very fast) and stream it to the user as DASH or HLS, then can be displayed client-side with https://github.com/kaltura/kaltura-player-js or videojs
* Kurento is a full-fledge media streaming server, it seems **very complicated**, but they have a demo for a media player client/server configuration: https://doc-kurento.readthedocs.io/en/latest/tutorials/java/tutorial-player.html

Have you considered any of those in the past? what is your opinion?

**Comment by junzhoum** [comment]
So, for this approach. We will need:


  1.  Create another VM as webRTC server. This server should not co-exist with current web server. Is it easy to retrieve MP4 files and m3u8 files from object store? We had quite some problem with AWS.
  2.  Ngnix should be a better option as it support HLS.

Jun

**Comment by zonca** [comment]
* yes, we should have a separate VM handling streaming
* what kind of problem did you have with MP4 files from S3? object store is basically the same so we would probably have the same issues
* good to know you prefer HLS
* is it fine if I use https://d3pazsmb3fddzx.cloudfront.net/USMC_100002_access.mp4 for testing? I'll load it to Jetstream object store

So I can spin up my own VM and check if I can make this work or not.

**Comment by junzhoum** [comment]
Andrea,

Thank you very much! I’ll forward you the error logs we have in the next email. It is fine you can use that mp4 file for testing.

Jun

**Comment by zonca** [comment]
ok, I have a first version which seems to be working,

* I have your mp4 video file which is hosted on Jetstream object store
* I have a medium VM on jetstream which runs Kaltura-NGINX and streams the mp4 into HLS
* the stream is at: https://js-170-203.jetstream-cloud.org/hls/swift/v1/videostream/USMC_100002_access.mp4/index.m3u8
* I can point a player to that and seems to work fine: [Videojs test](https://videojs-http-streaming.netlify.app/?debug=false&autoplay=false&muted=false&minified=false&sync-workers=false&liveui=false&partial=false&llhls=false&url=https%3A%2F%2Fjs-170-203.jetstream-cloud.org%2Fhls%2Fswift%2Fv1%2Fvideostream%2FUSMC_100002_access.mp4%2Findex.m3u8&type=application%2Fx-mpegURL&keysystems=&buffer-water=false&override-native=true&preload=auto&mirror-source=true)
* Or go to https://videojs-http-streaming.netlify.app/ and paste the url above

Please take a look and send me some feedback, I will work again on this in a couple of weeks.

**Comment by junzhoum** [comment]
Wow, this is wonderful! I am so impressed. Thank you very much!

So our website VM only need to integrate this player in the webpage?

Jun

**Comment by zonca** [comment]
correct, however this VM is under the ECSS account, and also it is better if you set it up yourself and make sure you can maintain it.

But for now let's make some testing with this VM.
Then if we decide this is the way to go, you can set it up yourself.

Of course this is a lot more risky than Cloudfront, here we are maintaining and managing the streaming server, which seems quite complex, best way is to test it for a few weeks and see how it goes.

Whenever I do development that is of generic interest, I write a tutorial, which is useful for documentation:

https://zonca.dev/2021/03/video-streaming-jetstream.html

**Comment by junzhoum** [comment]
You are right. I’ll setup my own VM and object container to give it a test. I guess VM on horizon should work as atmosphere is dying.

Thank you.

Jun

**Comment by zonca** [comment]
I think it would be better if we first do testing on my instance, before deploying another one, I can give you ssh root access to my instance if you email me your SSH public key.

**Comment by zonca** [comment]
please let me know if you wish to:

* do any testing on my VM (in case send me SSH key)
* suggest other test I can do on this setup

**Comment by junzhoum** [comment]
Hi Andrea,

Thanks for your email. I was contacting our student developer, but did not get his reply. I would like to include video samples in our website from your VM. Then we will see if more videos can be there. We are still evaluating AWS vs object storage.

Sorry for slow response. I’ll contact you when I have the update.

Thank you.

**Comment by zonca** [comment]
unless you plan to test the setup soon, I'll deactivate the VM next week to save resources.

**Comment by zonca** [comment]
ok, I shelved the image, so it doesn't burn through the ECSS allocation (not MIRC's allocation).
I can restore it later on if you want to test it.
